Source/Core/VideoBackends/D3D12/D3DUtil.cpp

```diff
@@ -0,0 +1,882 @@
+
+#include <cctype>
+#include <list>
+#include <string>
+
+#include "VideoBackends/D3D12/D3DBase.h"
+#include "VideoBackends/D3D12/D3DCommandListManager.h"
+#include "VideoBackends/D3D12/D3DDescriptorHeapManager.h"
+#include "VideoBackends/D3D12/D3DShader.h"
+#include "VideoBackends/D3D12/D3DState.h"
+#include "VideoBackends/D3D12/D3DTexture.h"
+#include "VideoBackends/D3D12/D3DUtil.h"
+#include "VideoBackends/D3D12/GeometryShaderCache.h"
+#include "VideoBackends/D3D12/PixelShaderCache.h"
+#include "VideoBackends/D3D12/VertexShaderCache.h"
+#include "VideoBackends/D3D12/Render.h"
+
+namespace DX12
+{
+
+extern D3D12_BLEND_DESC clearblendstates12[4];
+extern D3D12_DEPTH_STENCIL_DESC cleardepthstates12[3];
+extern D3D12_BLEND_DESC resetblendstate12;
+extern D3D12_DEPTH_STENCIL_DESC resetdepthstate12;
+extern D3D12_RASTERIZER_DESC resetraststate12;
+
+namespace D3D
+{
+	inline void ResourceBarrier(ID3D12GraphicsCommandList* commandList, ID3D12Resource* resource, D3D12_RESOURCE_STATES stateBefore, D3D12_RESOURCE_STATES stateAfter, UINT subresource)
+	{
+		if (stateBefore == stateAfter) return;
+
+		CHECK(resource, "NULL resource passed to ResourceBarrier.");
+
+		D3D12_RESOURCE_BARRIER resourceBarrierDesc = {
+			D3D12_RESOURCE_BARRIER_TYPE_TRANSITION, // D3D12_RESOURCE_TRANSITION_BARRIER_DESC Transition
+			D3D12_RESOURCE_BARRIER_FLAG_NONE,       // D3D12_RESOURCE_BARRIER_FLAGS Flags
+
+			// D3D12_RESOURCE_TRANSITION_BARRIER_DESC Transition
+			{
+				resource,    // ID3D12Resource *pResource;
+				subresource, // UINT Subresource;
+				stateBefore, // UINT StateBefore;
+				stateAfter   // UINT StateAfter;
+			}
+		};
+
+		commandList->ResourceBarrier(1, &resourceBarrierDesc);
+	}
+
+// Ring buffer class, shared between the draw* functions
+class UtilVertexBuffer
+{
+public:
+	UtilVertexBuffer(int size) : buf(nullptr), offset(0), max_size(size)
+	{
+		CheckHR(
+			device12->CreateCommittedResource(
+				&CD3DX12_HEAP_PROPERTIES(D3D12_HEAP_TYPE_UPLOAD),
+				D3D12_HEAP_FLAG_NONE,
+				&CD3DX12_RESOURCE_DESC::Buffer(max_size),
+				D3D12_RESOURCE_STATE_GENERIC_READ,
+				nullptr,
+				IID_PPV_ARGS(&buf12)
+				)
+			);
+
+		// Map buffer for CPU access upon creation.
+		// On D3D12, CPU-visible resources are persistently mapped.
+
+		CheckHR(buf12->Map(0, nullptr, &buf12data));
+	}
+	~UtilVertexBuffer()
+	{
+		D3D::commandListMgr->DestroyResourceAfterCurrentCommandListExecuted(buf12);
+	}
+
+	// returns vertex offset to the new data
+	int AppendData(void* data, int size, int vertex_size)
+	{
+		if (offset + size >= max_size)
+		{
+			// wrap buffer around and notify observers
+			offset = 0;
+		}
+
+		offset = ((offset+vertex_size-1)/vertex_size)*vertex_size; // align offset to vertex_size bytes
+
+		memcpy((u8*)buf12data + offset, data, size);
+
+		offset += size;
+		return (offset - size) / vertex_size;
+	}
+
+	inline ID3D11Buffer* &GetBuffer() { return buf; }
+	inline ID3D12Resource* &GetBuffer12() { return buf12; }
```

Comment:
Functions defined in the class body are inline by default, the keyword isn't necessary. Also the ampersand should be against the type along with the asterisk